fix(ntheory): Make ntheory functions thread-safe

oscarbenjamin opened the pull request:
#### References to other Issues or PRs

General issue gh-28239 but see specifically https://github.com/sympy/sympy/issues/28239#issuecomment-3652002429

Alternative approach to gh-28768 which uses a threading lock whereas here we use RCU for lock free updating of the sieve's internal state.

#### Brief description of what is fixed or changed

A number of functions in ntheory use global variables, and can fail under multithreading. This commit makes them thead-safe by only allowing atomic reassignment of the global state using the Read, Copy, Update (RCU) memory model.

https://en.wikipedia.org/wiki/Read-copy-update

Since module level globals or instance attributes are stored in dicts and dicts in Python use per-object locks for thread-safety we depend on the atomicity of setting an item in a dict to give an atomic update step and all mutations are done by making a local copy of the global state and only updating when the state is fully complete and correct.

The arrays in the Sieve class are now only extended with
```python
    _list = _list + array(...)
```
rather than
```python
    _list += array(...)
```
meaning that a new array is created rather than mutating the existing array in place. Note that at the C level, += still makes a copy of the buffer to the new memory location. The extra overhead here from not using += is just creating a new Python object that wraps the new memory buffer rather than reassigning the newly copied buffer inside the existing Python object.

Also, any method that depends on the array being a certain size should hold a local copy of the array e.g.
```python
    _list = self.extend(N)
    answer = _list[-1]
    self._list = _list
    return answer
```
This ensures that the array `_list` is not mutated by other threads (since they do not use `+=`) even though other threads may have reassigned `self._list` to a different array. It is possible that two threads could call extend at the same time but either way self._list will only ever be assigned some correct list of some size and every thread can depend on knowing that after `_list = self._list` or `_list = self.extend(N)` they have a local variable that is not going to be mutated by other threads.

With these changes the ntheory functions mostly pass tests under pytest-run-parallel e.g.
```python
    $ pip install pytest-run-parallel
    $ pytest --parallel-threads=4 sympy/ntheory
```
The above can be tested both with a GIL build of CPython and with a free-threaded build.

Some tests have been marked to be skipped under `--parallel-threads` since they use the warnings filter (catch_warnings) which does not work under multithreading or they use other global state such as a global RNG.

#### Other comments

There are still some tests that sometimes fail non-deterministically under multithreading specifically

- `test_qs_factor`
- `test_residue`
- `test_partition_rec`

In the case of test_residue it seems that some functions like _discrete_log_pollard_rho and _discrete_log_index_calculus use RNGs and sometimes fail depending on the RNG state. It is not clear if the non-deterministic failures should be expected behaviour for these functions given that they use RNGs so it is possible that just the test suite is making faulty assumptions about the RNG state rather than that there is a bug in the functions being tested.

For qs_factor there is potentially a bug to do with the factor cache not being thread-safe but I have not investigated this closely.

The test_partition_rec test fails sometimes because recurrence_memo does not work correctly under multithreading.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* ntheory
   * Some function in the ntheory module were made thread-safe. These functions are indirectly used by other top-level sympy functions such as factorial so this change prevents e.g. factorial from returning incorrect results when used under high contention multithreading.
<!-- END RELEASE NOTES -->

ngoldbaum commented on sympy/ntheory/tests/test_residue.py:
```diff
@@ -338,6 +339,7 @@ def test_binomial_p_pow():
         assert _binomial_mod_prime_power(n, m, prime, power) == binomials[m] % prime**power
 
 
+@pytest.mark.thread_unsafe(reason="warning filter broken under multithreading")
```
Over in NumPy we just deprecated our warning control utilities in favor of pytest and the standard library: https://github.com/numpy/numpy/pull/29550

Both are thread-safe in Python 3.14 on the free-threaded build and you can opt-in on the GIL-enabled build via a command-line flag: https://docs.python.org/3/whatsnew/3.14.html#concurrent-safe-warnings-control

You can also configure pytest-run-parallel to automatically mark functions using `warns_deprecated_sympy` as thread-unsafe: https://github.com/Quansight-Labs/pytest-run-parallel?tab=readme-ov-file#caveats

oscarbenjamin replied:
This is a hangover from sympy's old test runner but it does just use `catch_warnings` so it should work under the same conditions. For now perhaps it is preferable to use
```
pytest -X context_aware_warnings --parallel-threads=4 ...
```
Or perhaps it can be added in pytest.conf:
https://github.com/sympy/sympy/blob/1d7fcba31c84aff49e5b2df7b58119efe41431c4/pyproject.toml#L1-L3

ngoldbaum commented on sympy/ntheory/generate.py:
```diff
@@ -41,8 +44,47 @@ class Sieve:
+    # to exactly when and how it is accessed.
+    #
+    # The pattern used here is known as Read, Copy, Update (RCU) and allows
+    # fast read access without the use of locks for synchronisation.
```
I wrote a PR to document this better in the free-threading guide: https://github.com/Quansight-Labs/free-threaded-compatibility/pull/302. You can maybe link there once that is merged?

oscarbenjamin replied:
Yes, that looks good.

ngoldbaum commented on sympy/ntheory/generate.py:
```diff
@@ -217,11 +284,12 @@ def primerange(self, a, b=None):
+        _list = self._extend(b)
+        ai = bisect_left(_list, a)
+        bi = bisect_left(_list, b)
+        yield from _list[ai:bi]
```
The `_list` is append-only, right? So the fact that `_list` is possibly mutable during the iteration isn't a problem because there will only ever be new values appended after `bi`.

oscarbenjamin replied:
The `self._list` can be reset by the `_reset` method but it can also be made shorter by a parallel thread assigning a shorter list, basically the same as https://github.com/Quansight-Labs/free-threaded-compatibility/pull/302#discussion_r2632771021. We cannot depend on `self._list` being anything after our single atomic `_list = self._list` read apart from:

- It has length at least 6
- It holds valid data for its length.
- It is an object that is never mutated in place.

These are not sufficient to know that `self._list[ai:bi]` gives the slice we want so we need our local `_list`.

ngoldbaum reviewed (commented):
Awesome, this is a really nice pattern we should use more.

oscarbenjamin commented on sympy/ntheory/generate.py:
```diff
@@ -41,8 +44,47 @@ class Sieve:
+    _n: int
+    _list: _array[int]
+    _tlist: _array[int]
+    _mlist: _array[int]
```
Presumably it was originally a list and was at some point changed to an array. I don't consider it public so we could rename but probably best to leave as is right now.

ngoldbaum commented on sympy/ntheory/generate.py:
```diff
@@ -41,8 +44,47 @@ class Sieve:
+    #    self._list = new_list
+    # There should never be a situation where self._list does not hold entirely
+    # valid data e.g. because the array was extended but the correct values
+    # have not yet been inserted into the array.
```
Yikes! I'll check if I detect any races if I run the tests in this PR with a TSan interpreter...

oscarbenjamin commented on sympy/ntheory/generate.py:
```diff
@@ -41,8 +44,47 @@ class Sieve:
+    # The pattern used here is known as Read, Copy, Update (RCU) and allows
+    # fast read access without the use of locks for synchronisation.
+    #
+    # Nothing should be allowed to mutate any of these arrays in-place e.g.
```
We couldn't use `bytes` here given that is only allows values up to `255` or perhaps you mean that it could split the number across several bytes which would be unpleasant and inefficient compared to using array (e.g. bisect would not work any more).

Yes, it would be good if NumPy had immutable arrays.

oscarbenjamin committed:
fix(ntheory): make ntheory mostly thread-safe

Fixes the ntheory Sieve class to be thread-safe by using copy-on-write
(COW) for the sieve arrays.

Also fixes the handling of global variables in partitions_.py to be
thread-safe by assigning them atomically.

With these changes, most tests pass with pytest-run-parallel:

    pip install pytest-run-parallel
    pytest --parallel-threads=4 sympy/ntheory

There are still a couple failures, but those are caused by other
problems that should be fixed in future.

These tests still fail:

    PARALLEL FAILED sympy/ntheory/tests/test_partitions.py::test__partition_rec
        - assert 1575 == 231
    PARALLEL FAILED sympy/ntheory/tests/test_qs.py::test_qs_factor
        - assert 1 > 1

These failures happen under free-threading:

    PARALLEL FAILED sympy/ntheory/tests/test_continued_fraction.py::test_continued_fraction
        - sympy.core.facts.InconsistentAssumptions: {
    PARALLEL FAILED sympy/ntheory/tests/test_partitions.py::test__partition_rec
        - assert 3 == 5
    PARALLEL FAILED sympy/ntheory/tests/test_partitions.py::test__partition
        - AssertionError: assert [1, 1, 2, 3, 3, 5, ...] == [1, 1, 2, 3, 5, 7, ...]
    PARALLEL FAILED sympy/ntheory/tests/test_qs.py::test_qs_factor
        - assert 1 > 1
    PARALLEL FAILED sympy/ntheory/tests/test_residue.py::test_residue
        - Failed: DID NOT RAISE <class 'ValueError'>

oscarbenjamin commented:
Rebased after gh-28813 which adds the type annotations to `generate.py` which is the module where the changes here are happening. Basically the part where I added type annotations everywhere got factored out to a separate PR.

oscarbenjamin committed:
Add pytest-run-parallel to requirements-dev.txt
